File: R/R/pcvreg.R
#' Procrustes cross-validation for multivariate regression models
#'
#' @param X
#' matrix with predictors from the training set.
#' @param Y
#' vector with response values from the training set.
#' @param ncomp
#' number of components to use (more than the expected optimal number).
#' @param center
#' logical, to center or not the data sets
#' @param scale
#' logical, to scale or not the data sets
#' @param cv
#' which split method to use for cross-validation (see description of method `pcvpls()` for details).
#' @param funlist
#' list with functions for particular implementation
#' @param cv.scope
#' scope for center/scale operations inside CV loop: 'global' — using globally computed mean and std
#' or 'local' — recompute new for each local calibration set.
#'
#' @description
#' This is a generic method, use `pcvpls()` or `pcvpcr()` instead.
#'
#' @importFrom stats sd
pcvreg <- function(X, Y, ncomp = min(nrow(X) - 1, ncol(X), 30), cv = list("ven", 4),
   center = TRUE, scale = FALSE, funlist = list(), cv.scope = "global") {

   nRows <- nrow(X)
   nPred <- ncol(X)

   if (is.null(dim(Y))) {
      dim(Y) <- c(nRows, 1)
   }

   nResp <- ncol(Y)

   # keep names if any
   attrs <- attributes(X)

  # get indices for cross-validation
   ind <- pcvcrossval(cv, nRows, Y[, 1])

   # get number of segments
   nSeg <- max(ind)

   # correct maximum number of components
   max.ncomp <- min(nrow(X) - ceiling(nRows / nSeg) - 1, nPred)
   if (ncomp > max.ncomp) {
      ncomp <- max.ncomp
   }

   # compute global center and scale values for predictors
   mX <- rep(0, nPred)
   mY <- rep(0, nResp)
   if (center) {
      mX <- apply(X, 2, mean)
      mY <- apply(Y, 2, mean)
   }

   sX <- rep(1, nPred)
   sY <- rep(1, nResp)
   if (scale) {
      sX <- apply(X, 2, sd)
      sY <- apply(Y, 2, sd)
   }

   # autoscale globally
   X <- scale(X, center = mX, scale = sX)
   Y <- scale(Y, center = mY, scale = sY)
   m <- funlist$getglobalmodel(X, Y, ncomp)

   # prepare empty matrix for pseudo-validation set
   Xpv <- matrix(0, nRows, nPred)
   D <- matrix(0, nSeg, ncomp)

   # this will be used only for PLS case
   R <- array(0, dim = c(nSeg, nPred, ncomp))

   # loop for computing the PV set
   for (k in seq_len(nSeg)) {

      # split data to calibration and validation
      ind.c <- ind != k
      ind.k <- ind == k

      X.k <- X[ind.k, , drop = FALSE]
      X.c <- X[ind.c, , drop = FALSE]
      Y.c <- Y[ind.c, , drop = FALSE]

      # if cv.scope is local autoscale locally
      if (cv.scope == "local") {

         mXl <- rep(0, nPred)
         mYl <- rep(0, nResp)
         if (center) {
            mXl <- apply(X.c, 2, mean)
            mYl <- apply(Y.c, 2, mean)
         }

         sXl <- rep(1, nPred)
         sYl <- rep(1, nResp)
         if (scale) {
            sXl <- apply(X.c, 2, sd)
            sYl <- apply(Y.c, 2, sd)
         }

         X.c <- scale(X.c, center = mXl, scale = sXl)
         Y.c <- scale(Y.c, center = mYl, scale = sYl)
         X.k <- scale(X.k, center = mXl, scale = sXl)
      }

      # compute local model for current segment
      m.k <- funlist$getlocalmodel(X.c, Y.c, m)

      # compute explained part of PV-set for current segment
      pvres <- funlist$getxpv(m, m.k, X.k)
      Xpv.hat <- pvres$X
      D[k, ] <- diag(pvres$D)

      # in case of PLS regression we save R as well
      if (!is.null(pvres$R)) R[k, , ] <- pvres$R

      # compute the orthogonal part of PV-set
      Xpv.orth <- getxpvorth(funlist$getqk(X.k, m.k), X.k, m$Pi)

      # add the orthogonal part
      Xpv[ind.k, ] <- Xpv.hat + Xpv.orth
   }

   # uncenter and unscale the data for global scope
   Xpv <- sweep(Xpv, 2, sX, "*")
   Xpv <- sweep(Xpv, 2, mX, "+")

   # add initial attributes
   attributes(Xpv) <- attrs

   # add additional attribute with diagonal elements
   attr(Xpv, "D") <- D

   # in case of PLS regression we save R as well
   if (!is.null(pvres$R)) {
      attr(Xpv, "Rk") <- R
      attr(Xpv, "R") <- m$R
   }

   return(Xpv)
}

File: MATLAB/pcvreg.m
function [Xpv, D] = pcvreg(X, Y, nComp, Center, Scale, CV, funlist, CVScope)
%PCVREG Compute matrix with pseudo-validation set for Regression model
%
% Arguments
% ---------
% X         matrix with predictors from calibration set
% Y         matrix with responses from calibration set
% nComp     number of components for decomposition
% Center    logical, mean center data columns or not
% Scale     logical, standardize data columns or not
% CV        defines cross-validation splits pattern (see 'help crossval')
% CVScope   defines if centering/scaling must be done globally ('global')
% or locally ('local'). Default value is 'global'.
%
% !!! This is a generic function, use 'pcvpcr()' or 'pcvpls()' instead !!!
%

   if size(Y, 1) < size(Y, 2)
       Y = Y';
   end

   nRows = size(X, 1);
   nPred = size(X, 2);
   nResp = size(Y, 2);
   

   % compute indices for cross-validation
   ind = crossval(CV, nRows, Y);
   
   % compute number of segments
   nSeg = max(ind);

   % adjust number of components if necessary
   maxNComp = min([nRows - round(nRows / nSeg) - 1, nPred]);
   if nComp > maxNComp
      nComp = maxNComp;
   end

   % compute and save mean and standard deviation
   if Center
      mXg = mean(X);
      mYg = mean(Y);
   else
      mXg = zeros(1, nPred);
      mYg = zeros(1, nResp);
   end

   if Scale
      sXg = std(X);
      sYg = std(Y);
   else
      sXg = ones(1, nPred);
      sYg = ones(1, nResp);
   end

   % autoscale the calibration set and compute global model
   X = (X - mXg) ./ sXg;
   Y = (Y - mYg) ./ sYg;
   m = funlist.getglobalmodel(X, Y, nComp);

   % prepare empty matrix for pseudo-validation set and scaling
   % coefficients
   Xpv = zeros(nRows, nPred);
   D = zeros(nSeg, nComp);
   
   % loop for computing the PV set
   for k = 1:nSeg

      % split data to calibration and validation
      indc = ind ~= k;
      indk = ind == k;
      
      Xk = X(indk, :);
      Xc = X(indc, :);
      Yc = Y(indc, :);
      
      % autoscale data in case of local scope
      if CVScope == "local"
         
         if Center
            mX = mean(Xc);
            mY = mean(Yc);
         else
            mX = zeros(1, nPred);
            mY = zeros(1, nResp);
         end

         if Scale
            sX = std(Xc);
            sY = std(Yc);
         else
            sX = ones(1, nPred);
            sY = ones(1, nResp);
         end
         
         Xc = (Xc - mX) ./ sX;
         Yc = (Yc - mY) ./ sY;
         Xk = (Xk - mX) ./ sX;
      end
            
      % compute local model for current segment
      mk = funlist.getlocalmodel(Xc, Yc, m);

      % compute explained part of PV-set for current segment
      [Xpvhat, Dk] = funlist.getxpv(m, mk, Xk);
      D(k, :) = diag(Dk);

      % compute the orthogonal part of PV-set
      Xpvorth = getxpvorth(funlist.getqk(Xk, mk), Xk, m.Pi);

      % add the orthogonal part
      Xpv(indk, :) = Xpvhat + Xpvorth;
   end

   % uscenter and unscale the data for global scope
   Xpv = Xpv .* sXg + mXg;
end

